lib/Onnxifi/GlowOnnxManager.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "GlowOnnxManager.h"
+
+#include "Base.h"
+
+#include <mutex>
+
+namespace glow {
+namespace onnxifi {
+GlowOnnxManager &GlowOnnxManager::get() {
+  static GlowOnnxManager manager;
+  return manager;
+}
+
+void GlowOnnxManager::addBackendId(BackendIdPtr backendId) {
+  assert(!isValid(backendId));
+  {
+    std::lock_guard<std::mutex> lock(m_);
+    backendIds_.insert(backendId);
+  }
+  assert(isValid(backendId));
+}
+
+BackendPtr GlowOnnxManager::createBackend(BackendIdPtr backendId) {
+  assert(isValid(backendId));
+  BackendPtr backend = new Backend(backendId);
+  {
+    std::lock_guard<std::mutex> lock(m_);
+    backends_.insert(backend);
+  }
+  assert(isValid(backend));
+  return backend;
+}
+
+EventPtr GlowOnnxManager::createEvent() {
+  EventPtr event = new glow::onnxifi::Event();
+  {
+    std::lock_guard<std::mutex> lock(m_);
+    events_.insert(event);
+  }
+  assert(isValid(event));
+  return event;
+}
+
+GraphPtr GlowOnnxManager::createGraph(BackendPtr backend) {
+  assert(isValid(backend));
+  GraphPtr graph = new glow::onnxifi::Graph(backend);
+  {
+    std::lock_guard<std::mutex> lock(m_);
+    graphs_.insert(graph);
+  }
+  assert(isValid(graph));
```

Comment:
Like I pointed out above, this check is not necessary. The return value of `insert` should already tell you whether the graph is valid or not.

Reply:
Yeah it's true, on the one hand I like how this encapsulates the logic around checking validity in one place but given that there's duplicated locking involved here I think it's worth it do the thing you're suggesting.